test/extensions/transport_sockets/tls/cert_selectors/integration_test.cc

```diff
@@ -0,0 +1,346 @@
+    AssertionResult result = xds_connection_->waitForNewStream(*dispatcher_, xds_streams_.back());
+    RELEASE_ASSERT(result, result.message());
+    auto& xds_stream = *xds_streams_.back();
+    xds_stream.startGrpcStream();
+
+    envoy::service::discovery::v3::DeltaDiscoveryRequest delta_discovery_request;
+    AssertionResult result2 = xds_stream.waitForGrpcMessage(*dispatcher_, delta_discovery_request);
+    RELEASE_ASSERT(result2, result2.message());
+    EXPECT_EQ(1, delta_discovery_request.resource_names_subscribe().size())
+        << "Should be 1 resource in DELTA_GRPC";
+    EXPECT_EQ(cert, delta_discovery_request.resource_names_subscribe().at(0))
+        << "Secret name doesn't match in the request";
+    if (fail) {
+      removeSecret(xds_stream, cert);
+      return xds_stream;
+    }
+    envoy::service::discovery::v3::DeltaDiscoveryResponse discovery_response;
+    discovery_response.set_type_url(Config::TestTypeUrl::get().Secret);
+    auto* resource = discovery_response.add_resources();
+    resource->set_name(cert);
+    resource->mutable_resource()->PackFrom(getServerSecretRsa(cert));
+    xds_stream.sendGrpcMessage(discovery_response);
+    return xds_stream;
+  }
+
+  void removeSecret(FakeStream& xds_stream, absl::string_view cert) {
+    envoy::service::discovery::v3::DeltaDiscoveryResponse discovery_response;
+    discovery_response.set_type_url(Config::TestTypeUrl::get().Secret);
+    discovery_response.add_removed_resources(cert);
+    xds_stream.sendGrpcMessage(discovery_response);
+    xds_stream.finishGrpcStream(Grpc::Status::Ok);
+  }
+
+  void waitCertsRequested(uint32_t count) {
+    test_server_->waitForCounterEq(onDemandStat("cert_requested"), count,
+                                   TestUtility::DefaultTimeout, dispatcher_.get());
+  }
+
+  std::string onDemandStat(absl::string_view stat) {
+    return listenerStatPrefix(absl::StrCat("on_demand_secret.", stat));
+  }
+
+  std::vector<FakeStreamPtr> xds_streams_;
+};
+
+TEST_P(OnDemandIntegrationTest, BasicSuccessWithPrefetch) {
+  on_server_init_function_ = [&]() {
+    createXdsConnection();
+    waitSendSdsResponse("server");
+  };
+  setup(R"EOF(
+  certificate_mapper:
+    name: static-name
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.transport_sockets.tls.cert_mappers.static_name.v3.StaticName
+      name: server
+  prefetch_secret_names:
+  - server
+  )EOF");
+  auto conn = std::make_unique<ClientSslConnection>(*this);
+  conn->waitForUpstreamConnection();
+  conn->sendAndReceiveTlsData("hello", "world");
+  conn.reset();
+  EXPECT_EQ(1, test_server_->counter("sds.server.update_success")->value());
+  EXPECT_EQ(0, test_server_->counter("sds.server.update_rejected")->value());
+  EXPECT_EQ(1, test_server_->counter(onDemandStat("cert_requested"))->value());
+  EXPECT_EQ(1, test_server_->gauge(onDemandStat("cert_active"))->value());
+}
+
+TEST_P(OnDemandIntegrationTest, BasicSuccessWithoutPrefetch) {
```

Comment:
Can you please add a test case with extension function that provides the name of the secret to fetch?